tests/vm_delete_protection_test.go

```diff
@@ -0,0 +1,87 @@
+package tests
+
+import (
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/rand"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"kubevirt.io/ssp-operator/tests/env"
+)
+
+const deleteProtectionLabel = "kubevirt.io/vm-delete-protection"
+
+var _ = Describe("VM delete protection", func() {
+
+	var vm *kubevirtv1.VirtualMachine
+
+	BeforeEach(func() {
+		waitUntilDeployed()
+	})
+
+	AfterEach(func() {
+		err := apiClient.Get(ctx, client.ObjectKeyFromObject(vm), vm)
+		if !errors.IsNotFound(err) {
+			Eventually(func() error {
+				err := apiClient.Get(ctx, client.ObjectKeyFromObject(vm), vm)
+				if err != nil {
+					return err
+				}
+
+				vm.Labels[deleteProtectionLabel] = "False"
+				return apiClient.Update(ctx, vm)
+			}, env.ShortTimeout(), time.Second).Should(Succeed())
+
+			Expect(apiClient.Delete(ctx, vm)).To(Succeed())
+			waitForDeletion(client.ObjectKeyFromObject(vm), &kubevirtv1.VirtualMachine{})
+		}
+	})
+
+	It("should not allow to delete a VM if the protection is enabled", func() {
+		vm = createVMWithDeleteProtection("True")
+
+		Expect(apiClient.Delete(ctx, vm)).To(
+			MatchError(ContainSubstring("VirtualMachine %v cannot be deleted, remove delete protection", vm.Name)))
+	})
+
+	It("should be able to delete a VM if the protection is disabled", func() {
+		vm = createVMWithDeleteProtection("False")
+
+		Expect(apiClient.Delete(ctx, vm)).To(Succeed())
+		waitForDeletion(client.ObjectKeyFromObject(vm), &kubevirtv1.VirtualMachine{})
+	})
+
+	It("should be able to delete a VM if the VM does not have any label", func() {
+		vm = createVMWithDeleteProtection("True")
+
+		Eventually(func() error {
+			err := apiClient.Get(ctx, client.ObjectKeyFromObject(vm), vm)
+			if err != nil {
+				return err
+			}
+			vm.Labels = nil
+			return apiClient.Update(ctx, vm)
+		}, env.ShortTimeout(), time.Second).Should(Succeed())
```

Comment:
Can't you simply create a VM without the label in the first place?

Reply:
The idea behind this test is to check the corner case that the VMs does not have any labels at all `labels == nil`. I can create the VM without the `kubevirt.io/vm-delete-protection`, but it does not guarantee that the VM has some default label, e.g., `kubevirt.io/vm`.

Reply:
Changed to create the VM without any labels.